src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -1843,22 +1862,45 @@ private void EnsureRootBoundForNullabilityIfNecessary()
 
             Debug.Assert(Root == GetBindableSyntaxNode(Root));
 
-            // https://github.com/dotnet/roslyn/issues/35037: Speculative models are going to have to do something more advanced
-            // here. They will need to run nullable analysis up to the point that is being speculated on, and
-            // then take that state and run analysis on the statement or expression being speculated on.
-            // Currently, it will return incorrect info because it's just running analysis on the speculated
-            // part.
             var binder = GetEnclosingBinder(GetAdjustedNodePosition(bindableRoot));
             var boundRoot = Bind(binder, bindableRoot, diagnostics);
-            boundRoot = RewriteNullableBoundNodesWithSnapshots(boundRoot, binder, diagnostics, out _);
+            if (IsSpeculativeSemanticModel)
+            {
+                ensureSpeculativeNodeBound();
+            }
+            else
+            {
+                bindAndRewrite(takeSnapshots: true);
+            }
 
-            if (Compilation.NullableAnalysisEnabled && !IsSpeculativeSemanticModel)
+            void ensureSpeculativeNodeBound()
             {
+                // Not all speculative models are created with existing snapshots. Attributes,
+                // TypeSyntaxes, and MethodBodies do not depend on existing state in a member,
+                // and so the SnapshotManager can be null in these cases.
+                if (_lazyGuardedSnapshotManager is null)
+                {
+                    bindAndRewrite(takeSnapshots: false);
+                    return;
+                }
+
+                var analyzedNullabilitiesBuilder = ImmutableDictionary.CreateBuilder<BoundExpression, (NullabilityInfo, TypeSymbol)>();
+                boundRoot = NullableWalker.AnalyzeAndRewriteSpeculation(_speculatedPosition, boundRoot, binder, _lazyGuardedSnapshotManager);
                 GuardedAddBoundTreeForStandaloneSyntax(bindableRoot, boundRoot);
             }
+
+            void bindAndRewrite(bool takeSnapshots)
+            {
+                boundRoot = RewriteNullableBoundNodesWithSnapshots(boundRoot, binder, diagnostics, takeSnapshots, out var snapshotManager);
+#if DEBUG
+                // Don't actually cache the results if the nullable analysis is not enabled in debug mode.
```

Comment:
>// Don't actually cache the results if the nullable analysis is not enabled in debug mode. [](start = 16, length = 90)

What are we trying to achieve by doing this?

Reply:
We're trying to run the analysis on more tests, like we do with the IOperation test hook.

---
In reply to: [295524754](https://github.com/dotnet/roslyn/pull/36563#discussion_r295524754) [](ancestors = 295524754)